### ISSUE TITLE
Rescue block (without a begin) within define_method throws a parser error.

### ISSUE THREAD

**Issue by EdwardDiehl** [DamirSvrtan/fasterer#54]
This is a source code from my rails app models concern

```
# frozen_string_literal: true

module RecordNotUniqueHandler
  extend ActiveSupport::Concern

  included do
    define_method :create_or_update do |*args|
      super(*args)

    rescue ActiveRecord::RecordNotUnique => e
      attr_name, *scope = e.message
                           .match(/Key \(([^\)]+)\)/)[1].split(/,\s*/)

      message = +(errors.generate_message(attr_name, :taken))
      message << " (scope: #{scope.to_sentence})" if scope.any?
      errors.add attr_name, message
      false
    end
  end
end
```

fasterer outputs this error in the console
```
Fasterer was unable to process some files because the
internal parser is not able to read some characters or
has timed out. Unprocessable files were:
-----------------------------------------------------
app/models/concerns/record_not_unique_handler.rb - Racc::ParseError - (string):10 :: parse error on value ["rescue", 10] (kRESCUE)
```

Looks like it fails on this line `rescue ActiveRecord::RecordNotUnique => e`

**Comment by DamirSvrtan** [comment]
Hi @EdwardDiehl!

The reason why fasterer cannot process this file is because it has a syntax error. Just try running it with ruby and you will get an error:

```shell
concern.rb:10: syntax error, unexpected keyword_rescue, expecting keyword_end
    rescue ActiveRecord::RecordNotUnique => e
          ^
concern.rb:10: syntax error, unexpected =>
    rescue ActiveRecord::RecordNotUnique => e
```

The problem is that you are using the rescue without begin-rescue-end style syntax and that is only supported with regular `def`s, now with the define_method.

What you can do is either define it like this:

```ruby
module RecordNotUniqueHandler
  extend ActiveSupport::Concern

  included do
    def create_or_update(*args)
      super(*args)

    rescue ActiveRecord::RecordNotUnique => e
      attr_name, *scope = e.message
                           .match(/Key \(([^\)]+)\)/)[1].split(/,\s*/)

      message = +(errors.generate_message(attr_name, :taken))
      message << " (scope: #{scope.to_sentence})" if scope.any?
      errors.add attr_name, message
      false
    end
  end
end
```
 or use the begin-end block to accomplish the same thing with define_method:

```ruby
# frozen_string_literal: true

module RecordNotUniqueHandler
  extend ActiveSupport::Concern

  included do
    define_method :create_or_update do |*args|
      begin
        super(*args)
      rescue ActiveRecord::RecordNotUnique => e
        attr_name, *scope = e.message
                             .match(/Key \(([^\)]+)\)/)[1].split(/,\s*/)

        message = +(errors.generate_message(attr_name, :taken))
        message << " (scope: #{scope.to_sentence})" if scope.any?
        errors.add attr_name, message
        false
      end
    end
  end
end
```

Closing as this is not a fasterer issue.

**Comment by graywolf** [comment]
I think you don't need the `begin` since ruby 2.5:

```
+   $ cat test.rb
module Foo
  def self.included(klass)
    pp 'xxxx'
    define_method :create_or_update do |*args|
      true
    rescue => e
      false
    end
  end
end

class Bar
  include Foo
end

+   $ ruby test.rb
"xxxx"

+   $ ruby --version
ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-linux]
```

(My version is without active record since I don't use that but should not be
relevant to this issue.)

I dunno, it's possible I'm missing something obvious here...

**Comment by DamirSvrtan** [comment]
Hey @graywolf - you are using the begin-less syntax again in a method definition -> in `def.self.included`. That's why it's working.

**Comment by graywolf** [comment]
Ok so going with original code (I've installed rails in docker to try it out):

```
+[dockwolf@test devel]$ cat test.rb 
# frozen_string_literal: true

require 'active_support'

module RecordNotUniqueHandler
  extend ActiveSupport::Concern

  included do
    puts "Executed"
    define_method :create_or_update do |*args|
      super(*args)
    rescue ActiveRecord::RecordNotUnique => e
      attr_name, *scope = e.message
                           .match(/Key \(([^\)]+)\)/)[1].split(/,\s*/)

      message = +(errors.generate_message(attr_name, :taken))
      message << " (scope: #{scope.to_sentence})" if scope.any?
      errors.add attr_name, message
      false
    end
  end
end

class Foo
  include RecordNotUniqueHandler
end
+[dockwolf@test devel]$ ruby test.rb 
Executed
```

so it parses correctly... how should I run the file to get the error you have shown above?

Fasterer indeed fails to parse the file

```
+[dockwolf@test devel]$ fasterer test.rb 
Fasterer was unable to process some files because the
internal parser is not able to read some characters or
has timed out. Unprocessable files were:
-----------------------------------------------------
test.rb - Racc::ParseError - (string):12 :: parse error on value ["rescue", 12] (kRESCUE)

1 file inspected, 0 offenses detected, 1 unparsable file found
```

however as I've said above, I have issues reproducing the issue in ruby...

**Comment by graywolf** [comment]
Any update on this?

**Comment by DamirSvrtan** [comment]
@graywolf which version of Ruby are you using?

**Comment by graywolf** [comment]
The previous attempt was in `2.5.3`, this time with newer one:

```
+[dockwolf@test devel]$ ruby --version
ruby 2.6.0p0 (2018-12-25 revision 66547) [x86_64-linux]
+[dockwolf@test devel]$ cat test.rb
# frozen_string_literal: true

require 'active_support'

module RecordNotUniqueHandler
  extend ActiveSupport::Concern

  included do
    puts "Executed"
    define_method :create_or_update do |*args|
      super(*args)
    rescue ActiveRecord::RecordNotUnique => e
      attr_name, *scope = e.message
                           .match(/Key \(([^\)]+)\)/)[1].split(/,\s*/)

      message = +(errors.generate_message(attr_name, :taken))
      message << " (scope: #{scope.to_sentence})" if scope.any?
      errors.add attr_name, message
      false
    end
  end
end

class Foo
  include RecordNotUniqueHandler
end
+[dockwolf@test devel]$ ruby test.rb
Executed
+[dockwolf@test devel]$ fasterer test.rb
NOTE: RubyParser::V26 undefined, using RubyParser::V25.
Fasterer was unable to process some files because the
internal parser is not able to read some characters or
has timed out. Unprocessable files were:
-----------------------------------------------------
test.rb - Racc::ParseError - (string):12 :: parse error on value ["rescue", 12] (kRESCUE)

1 file inspected, 0 offenses detected, 1 unparsable file found
```

**Comment by graywolf** [comment]
fasterer uses ruby_parser right? Notice that while ruby_parser considers `test1.rb` invalid, ruby executes correctly both of them:

```
+[dockwolf@test devel]$ cat test1.rb
define_method :bar  do
	puts "bar"
rescue
end
bar
+[dockwolf@test devel]$ cat test2.rb
define_method(:bar)  do
	puts "bar"
rescue
end
bar
+[dockwolf@test devel]$ ruby test1.rb
bar
+[dockwolf@test devel]$ ruby test2.rb
bar
+[dockwolf@test devel]$ ruby_parse test1.rb
# file = test1.rb loc = 5
/usr/lib/ruby/2.6.0/racc/parser.rb:528:in `on_error' #<Racc::ParseError: test1.rb:3 :: parse error on value ["rescue", 3] (kRESCUE)> for test1.rb
  /usr/lib/ruby/2.6.0/racc/parser.rb:528:in `on_error'
  /home/dockwolf/.gem/ruby/2.6.0/gems/ruby_parser-3.12.0/lib/ruby_parser_extras.rb:1175:in `on_error'
  /usr/lib/ruby/2.6.0/racc/parser.rb:259:in `_racc_do_parse_c'
  /usr/lib/ruby/2.6.0/racc/parser.rb:259:in `do_parse'
  /home/dockwolf/.gem/ruby/2.6.0/gems/ruby_parser-3.12.0/lib/ruby_parser_extras.rb:1082:in `block in process'
  /usr/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
  /usr/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
  /usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
  /usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
  /usr/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
  /home/dockwolf/.gem/ruby/2.6.0/gems/ruby_parser-3.12.0/lib/ruby_parser_extras.rb:1070:in `process'
  /home/dockwolf/.gem/ruby/2.6.0/gems/ruby_parser-3.12.0/lib/ruby_parser.rb:35:in `block in process'
  /home/dockwolf/.gem/ruby/2.6.0/gems/ruby_parser-3.12.0/lib/ruby_parser.rb:32:in `each'
  /home/dockwolf/.gem/ruby/2.6.0/gems/ruby_parser-3.12.0/lib/ruby_parser.rb:32:in `process'
  /home/dockwolf/.gem/ruby/2.6.0/gems/ruby_parser-3.12.0/bin/ruby_parse:48:in `block in <top (required)>'
  /home/dockwolf/.gem/ruby/2.6.0/gems/ruby_parser-3.12.0/bin/ruby_parse:29:in `each'
  /home/dockwolf/.gem/ruby/2.6.0/gems/ruby_parser-3.12.0/bin/ruby_parse:29:in `<top (required)>'
  /home/dockwolf/.gem/ruby/2.6.0/bin/ruby_parse:23:in `load'
  /home/dockwolf/.gem/ruby/2.6.0/bin/ruby_parse:23:in `<main>'
done



 0.00s:  1439.96 l/s:   14.06 Kb/s:    0 Kb:    5 loc:TOTAL
+[dockwolf@test devel]$ ruby_parse test2.rb
# file = test2.rb loc = 5
s(:block,
 s(:iter,
  s(:call, nil, :define_method, s(:lit, :bar)),
  0,
  s(:rescue,
   s(:call, nil, :puts, s(:str, "bar")),
   s(:resbody, s(:array), nil))),
 s(:call, nil, :bar))
done



 0.00s:  2258.94 l/s:   22.50 Kb/s:    0 Kb:    5 loc:TOTAL
```

**Comment by DamirSvrtan** [comment]
Hi @graywolf, I've raised an issue on Ruby parser (which is what fasterer uses under the hood to parse ruby). Thank you for reporting the version as well bc I was not able to previously reproduce it with ruby 2.3 (nor 2.4 for that matter).

We have to wait until the ruby parser is updated. Is this blocking you in your development cycle?

**Comment by EdwardDiehl** [comment]
Thank you guys for reproducing the issue and thank you @DamirSvrtan for the fasterer gem. Let's wait for the new version of ruby parser gem. It is not blocking issue for me.

**Comment by monfresh** [comment]
In my app, ruby_parser version 3.12.0 fixes the issue. However, I can't update to 3.12.0 because `fasterer` is being too strict about the versioning: `~> 3.11.0`, which will only allow `3.11.x` versions. @DamirSvrtan is it possible to relax the version constraint please, or update it to 3.12.0?

**Comment by DamirSvrtan** [comment]
@monfresh I've released a new version of [Fasterer](https://rubygems.org/gems/fasterer) with looser version constraints from now on, so we don't hit the same issues.

However, this does not solve the problem on Ruby 2.5, at least from what I'm testing out. @EdwardDiehl what about you?

**Comment by monfresh** [comment]
Thanks @DamirSvrtan! That worked for me. In my case, I'm not using `define_method`. ruby_parser 3.11.0 was failing to parse something like this:
```ruby
def some_method
  enumerable.each do |element|
    do_something_that_might_raise_an_error
  rescue StandardError => e
    handle_the_error
  end
end
```
After upgrading, it is able to parse it.

**Comment by EdwardDiehl** [comment]
> However, this does not solve the problem on Ruby 2.5, at least from what I'm testing out. 

@DamirSvrtan 
The problem is still here for me as well
`gem 'fasterer', '~> 0.4.2'`
`ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-linux]`

It still outputs `Racc::ParseError`

**Comment by DamirSvrtan** [comment]
The ruby_parser team has released support for Ruby 2.6.0 today so I have released a new version of fasterer that supports 2.6 as well. it's on rubygems, version `0.5.1`. 

Thank you all for collaborating on this and pointing me in the right direction!
